src/LiveComponent/src/EventListener/LiveComponentSubscriber.php

```diff
@@ -107,6 +112,16 @@ public function onKernelRequest(RequestEvent $event): void
             throw new BadRequestHttpException('Invalid CSRF token.');
         }
 
+        if ('_batch' === $action) {
+            // use batch controller
+            $request->attributes->set('_controller', 'ux.live_component.batch_action_controller');
+            $request->attributes->set('_component_service_id', $metadata->getServiceId());
+
+            $this->mountComponent($this->container->get(ComponentFactory::class)->get($componentName), $request);
+
+            return;
+        }
```

Comment:
Ok, I see the complexity. I think the flow might look simpler if we furthered your idea more: try to move as much as possible into `onKernelRequest()`. Looking at the stuff in `onKernelController()`, I'm wondering if we could make `onKernelRequest()` generally look like this:

A) Fetch the action and component names: https://github.com/weaverryan/ux/blob/5738d7751684ff9d2e64679fca3fd48668f73d90/src/LiveComponent/src/EventListener/LiveComponentSubscriber.php#L78-L79

B) Fetch the `$metadata` and do the checks: https://github.com/weaverryan/ux/blob/5738d7751684ff9d2e64679fca3fd48668f73d90/src/LiveComponent/src/EventListener/LiveComponentSubscriber.php#L83-L92

C) If `$request->attributes->get('_mounted_component')`, then get it. And SKIP step `C1-OPTIONAL`. Though, I'm not sure if this would ever be set in practice (see below: in the case of a sub-request, `_controller` would also be set, so this entire method would be skipped, including this part).

C-OPTIONAL) If there is not a `_mounted_component` yet, then fetch the component from the container, mount it, and set it onto `_mounted_component`. 

D) Check whether the given `$action` is allowed for this component https://github.com/weaverryan/ux/blob/5738d7751684ff9d2e64679fca3fd48668f73d90/src/LiveComponent/src/EventListener/LiveComponentSubscriber.php#L146-L152

E) If there is a `X-CSRF-TOKEN` header, check CSRF (I don't think we need to do this on POST only, even if that's when it will happen in practice)

F) Set `_controller` into request attributes. This is where we'll have special cases for the default action, `_batch` and the normal situation. Basically this chunk - https://github.com/weaverryan/ux/blob/5738d7751684ff9d2e64679fca3fd48668f73d90/src/LiveComponent/src/EventListener/LiveComponentSubscriber.php#L94-L102 - and this chunk - https://github.com/weaverryan/ux/blob/5738d7751684ff9d2e64679fca3fd48668f73d90/src/LiveComponent/src/EventListener/LiveComponentSubscriber.php#L115-L125

This means that, by the end of `onKernelRequest` we have (in a very consistent way), mounted the component and set the controller. ALL of this would be skipped if `_controller` is already set (i.e. the sub-request). 

In `onKernelView()`, I think we only need 2 things:

A) For the sub-request, change the `_controller` to trick the system - https://github.com/weaverryan/ux/blob/5738d7751684ff9d2e64679fca3fd48668f73d90/src/LiveComponent/src/EventListener/LiveComponentSubscriber.php#L154-L156 - but I don't think we should use `_mounted_component... I think it's clearer if this is always set early and is always, simply, the "mounted component". Instead, we should send some second flag down into the sub-request like `_set_to_component_action` `true`
B) The `LiveArg` handling.

Let me know if this makes sense - I'm kind of "arm-chair" coding, sorry. But I think you were on the right path when you were thinking about moving nearly everything into `onKernelRequest`. We should be able to get a consistent flow in that method of (on a high level) 1) mount component 2) set controller. And on a sub-request, both of these would already have occurred, so we would bail out early.

Reply:
Hah, this is a lot to process!

One thing we need to do is _not_ set `_mounted_component` in the request listener for non-batch-requests. This would double-instantiate the component. We have to do this for batch requests but not for non-batch. I might be wrong, but I think this is what you're suggesting in C/C-Optional.